pkg/rulefmt/rulefmt.go

```diff
@@ -58,10 +60,50 @@ type RuleGroups struct {
 	Groups []RuleGroup `yaml:"groups"`
 }
 
+// RuleGroup is a list of sequentially evaluated recording and alerting rules.
+type RuleGroup struct {
+	Name     string         `yaml:"name"`
+	Interval model.Duration `yaml:"interval,omitempty"`
+	Rules    []RuleNode     `yaml:"rules"`
+}
+
+// RuleNode adds yaml.v3 layer to support line and column outputs for invalid rules.
+type RuleNode struct {
+	Record      yaml.Node         `yaml:"record,omitempty"`
+	Alert       yaml.Node         `yaml:"alert,omitempty"`
+	Expr        yaml.Node         `yaml:"expr"`
+	For         model.Duration    `yaml:"for,omitempty"`
+	Labels      map[string]string `yaml:"labels,omitempty"`
+	Annotations map[string]string `yaml:"annotations,omitempty"`
+}
+
+// ruleGroups type of yaml.v3 features in the groups field.
 type ruleGroups struct {
 	Groups []yaml.Node `yaml:"groups"`
 }
 
+// validateRules type for validating the groups inside the
+// rules files under strict mode, supported under yaml.v2
+type validateRules struct {
+	Groups []validateGroup `yaml:"groups"`
+}
+
+type validateGroup struct {
+	Name     string         `yaml:"name"`
+	Interval model.Duration `yaml:"interval,omitempty"`
+	Rules    []Rule         `yaml:"rules"`
+}
+
+// Rule describes an alerting or recording rule.
+type Rule struct {
+	Record      string            `yaml:"record,omitempty"`
+	Alert       string            `yaml:"alert,omitempty"`
+	Expr        string            `yaml:"expr"`
+	For         model.Duration    `yaml:"for,omitempty"`
+	Labels      map[string]string `yaml:"labels,omitempty"`
+	Annotations map[string]string `yaml:"annotations,omitempty"`
+}
+
 // Validate validates all rules in the rule groups.
 func (g *RuleGroups) Validate(node ruleGroups) (errs []error) {
```

Comment:
Now that we're here, can we also fix this function? It is a public function but it is expecting a private type.

Reply:
Which would fix https://github.com/prometheus/prometheus/issues/7128

Reply:
Sure. I will update this to public. however, the merge might take a good amount of time due to lack of sufficient features in yaml v3.

Reply:
Done. The function is public now.